asmcomp/emitaux.ml

```diff
@@ -238,24 +238,65 @@ let emit_frames a =
     a.efa_def_label lbl;
     a.efa_string name
   in
-  let emit_defname (_filename, defname) (file_lbl, lbl) =
+  let emit_defname (_filename, defname, loc) (file_lbl, lbl) =
+    let emit_loc (chars, char_end_offset) =
+      a.efa_32 chars;
+      a.efa_32 char_end_offset
+    in
     (* These must be 32-bit aligned, both because they contain a
        32-bit value, and because emit_debuginfo assumes the low 2 bits
        of their addresses are 0. *)
     a.efa_align 4;
     a.efa_def_label lbl;
     a.efa_label_rel file_lbl 0l;
+    (* Include the additional 64-bits of location information which didn't pack
+       in the main 64-bit word *)
+    Option.iter emit_loc loc;
     a.efa_string defname
   in
-  let pack_info fd_raise d has_next =
-    let line = min 0xFFFFF d.Debuginfo.dinfo_line
-    and char_start = min 0xFF d.Debuginfo.dinfo_char_start
-    and char_end = min 0x3FF d.Debuginfo.dinfo_char_end
+  let fully_pack_info fd_raise d has_next =
+    (* Fully packed debuginfo:
+       0lllllmaaabbboooooddddddddddddkn
+         0           - d points to a name_info struct
+         l (12 bits) - start line number
+         m (3 bits)  - offset of end line number from start
+         a (6 bits)  - start character offset relative to start line
+         b (7 bits)  - end character offset relative to end line
+         o (9 bits)  - difference between start bol and end bol
+                       (so o + b is the end character relative to start line)
+         d (24 bits) - memory offset to name_info struct
+         k (1 bit)   - fd_raise flag
+         n (1 bit)   - has_next flag *)
+    let open Debuginfo in
+    let kind = if fd_raise then 1 else 0
+    and has_next = if has_next then 1 else 0
+    and char_end = d.dinfo_char_end + d.dinfo_start_bol - d.dinfo_end_bol in
+    let char_end_offset = d.dinfo_end_bol - d.dinfo_start_bol in
+    Int64.(add (shift_left (of_int d.dinfo_line) 51)
+             (add (shift_left (of_int (d.dinfo_end_line - d.dinfo_line)) 48)
+                (add (shift_left (of_int d.dinfo_char_start) 42)
+                   (add (shift_left (of_int char_end) 35)
+                      (add (shift_left (of_int char_end_offset) 26)
+                         (add (shift_left (of_int kind) 1)
+                            (of_int has_next)))))))
```

Comment:
Note: to me this stuff would be easier to read using pipes.

```ocaml
shift_left (of_int d.dinfo_line) 51
|> add (shift_left (of_int (d.dinfo_end_line - d.dinfo_line)) 48)
|> ...
```

or using a list

```ocaml
List.fold_left (fun n (i, p) -> Int64.(add n (shift_left (of_int i) p))) 0L [
  (d.dinfo_line, 51);
  (d.dinfo_end_line - d.dinfo_line, 48);
  ...
]
```

Reply:
I agree (I'd use the pipes): I kept the style of what was there before (also knowing that there's a least one backend developer with much, much stronger opinions on these things than me!)